File: src/main/java/net/mcbrawls/sgui/ParentedGui.java
package net.mcbrawls.sgui;

import eu.pb4.sgui.api.gui.GuiInterface;
import eu.pb4.sgui.api.gui.layered.LayeredGui;
import org.jetbrains.annotations.Nullable;

public interface ParentedGui {
    /**
     * The parent of this GUI.
     */
    @Nullable
    default GuiInterface getParent() {
        throw new AssertionError();
    }

    /**
     * Sets the parent of this GUI.
     */
    default void setParent(@Nullable GuiInterface parent) {
        throw new AssertionError();
    }

    /**
     * Opens the parent GUI.
     * @return whether the gui was opened
     */
    default boolean openParent() {
        GuiInterface parent = this.getParent();
        if (parent == null) {
            return false;
        }

        if (parent.open()) {
            if (parent instanceof LayeredGui layeredGui) {
                layeredGui.markDirty();
            }

            return true;
        }

        return false;
    }

    /**
     * Gets the top-most parent in the parent heirarchy.
     */
    @Nullable
    default GuiInterface getFirstParent() {
        GuiInterface parent = this.getParent();

        if (parent == null) {
            return null;
        }

        while (true) {
            if (parent instanceof ParentedGui parentedGui) {
                GuiInterface newParent = parentedGui.getParent();
                if (newParent != null) {
                    parent = newParent;
                } else {
                    break;
                }
            } else {
                break;
            }
        }

        return parent;
    }

    /**
     * Whether this gui has a parent.
     */
    default boolean hasParent() {
        return this.getParent() != null;
    }
}

File: src/main/kotlin/net/mcbrawls/sgui/ParentedGui.kt
package net.mcbrawls.sgui

import eu.pb4.sgui.api.gui.GuiInterface
import eu.pb4.sgui.api.gui.layered.LayeredGui

interface ParentedGui {
    /**
     * The parent of this GUI.
     */
    var parent: GuiInterface?

    /**
     * Opens the parent GUI.
     * @return whether the gui was opened
     */
    fun openParent(): Boolean {
        val parent = parent ?: return false
        (parent as? LayeredGui)?.markDirty()
        return parent.open()
    }

    companion object {
        /**
         * Sets this GUI's parent GUI.
         */
        fun GuiInterface.setGuiParent(parent: GuiInterface) : GuiInterface {
            val parented = this as? ParentedGui ?: return this
            parented.parent = parent
            return this
        }

        /**
         * Gets this GUI's parent GUI.
         */
        fun GuiInterface.getGuiParent() : GuiInterface? {
            val parented = this as? ParentedGui ?: return null
            return parented.parent
        }

        /**
         * Gets the top-most parent in the parent heirarchy.
         */
        fun GuiInterface.getFirstParent(): GuiInterface? {
            var parent = getGuiParent() ?: return null

            while (true) {
                parent = parent.getGuiParent() ?: break
            }

            return parent
        }

        /**
         * Whether this gui has a parent.
         */
        fun GuiInterface.hasParent(): Boolean {
            return getGuiParent() != null
        }

        /**
         * Opens a GUI's parent GUI.
         */
        fun GuiInterface.openGuiParent(): GuiInterface {
            val parented = this as? ParentedGui ?: return this
            parented.openParent()
            return this
        }
    }
}
